crates/bevy_ecs/src/query/cache.rs

```diff
@@ -0,0 +1,475 @@
+use bevy_ecs::component::ComponentId;
+use bevy_ecs::prelude::World;
+use bevy_ecs::query::state::StorageId;
+use bevy_ecs::query::{FilteredAccess, QueryBuilder, QueryData, QueryFilter};
+use bevy_ecs::storage::{SparseSetIndex, TableId};
+use bevy_ecs::world::unsafe_world_cell::UnsafeWorldCell;
+use bevy_ecs::world::WorldId;
+use core::fmt::{Debug, Formatter};
+use fixedbitset::FixedBitSet;
+use log::warn;
+
+/// Borrow-only view over the non-cache fields of a QueryState.
+#[doc(hidden)]
+pub struct UncachedQueryBorrow<'a, D: QueryData, F: QueryFilter> {
+    pub(crate) world_id: WorldId,
+    pub(crate) component_access: &'a FilteredAccess,
+    pub(crate) fetch_state: &'a D::State,
+    pub(crate) filter_state: &'a F::State,
+}
+
+impl<'a, D: QueryData, F: QueryFilter> UncachedQueryBorrow<'a, D, F> {
+    /// # Panics
+    ///
+    /// If `world_id` does not match the [`World`] used to call `QueryState::new` for this instance.
+    ///
+    /// Many unsafe query methods require the world to match for soundness. This function is the easiest
+    /// way of ensuring that it matches.
+    #[inline]
+    #[track_caller]
+    pub fn validate_world(&self, world_id: WorldId) {
+        #[inline(never)]
+        #[track_caller]
+        #[cold]
+        fn panic_mismatched(this: WorldId, other: WorldId) -> ! {
+            panic!("Encountered a mismatched World. This QueryState was created from {this:?}, but a method was called using {other:?}.");
+        }
+
+        if self.world_id != world_id {
+            panic_mismatched(self.world_id, world_id);
+        }
+    }
+
+    /// Returns `true` if this query matches a set of components. Otherwise, returns `false`.
+    ///
+    /// # Safety
+    /// `archetype` must be from the `World` this [`UncachedQueryBorrow`] was initialized from.
+    pub unsafe fn matches_archetype(&self, archetype: &Archetype) -> bool {
+        D::matches_component_set(self.fetch_state, &|id| archetype.contains(id))
+            && F::matches_component_set(self.filter_state, &|id| archetype.contains(id))
+            && self.component_access.filter_sets.iter().any(|set| {
+                set.with
+                    .ones()
+                    .all(|index| archetype.contains(ComponentId::get_sparse_set_index(index)))
+                    && set
+                        .without
+                        .ones()
+                        .all(|index| !archetype.contains(ComponentId::get_sparse_set_index(index)))
+            })
+    }
+
+    /// Iterate through all new archetypes more recent than the provided [`ArchetypeGeneration`],
+    /// and call `f` on each of them.
+    pub fn iter_archetypes(
+        &self,
+        archetype_generation: ArchetypeGeneration,
+        archetypes: &Archetypes,
+        mut f: impl FnMut(&Archetype),
+    ) {
+        if self.component_access.required.is_empty() {
+            archetypes[archetype_generation..]
+                .iter()
+                .for_each(|archetype| {
```

Comment:
for uncached queries we're still iterating over all the archetypes here. Feels like we could do better with come component -> archetype caches, but that could be in a follow up